src/Membership/Voter/MembershipVoter.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace AppBundle\Membership\Voter;
+
+use AppBundle\Membership\MembershipPermissions;
+use AppBundle\Entity\Adherent;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\VoterInterface;
+
+class MembershipVoter implements VoterInterface
+{
+    public function vote(TokenInterface $token, $subject, array $attributes)
+    {
+        $adherent = $token->getUser();
+
+        if (null !== $subject || !$adherent instanceof Adherent) {
+            return self::ACCESS_ABSTAIN;
+        }
+
+        if (!in_array(MembershipPermissions::UNREGISTER, $attributes, true)) {
+            return self::ACCESS_ABSTAIN;
+        }
+
+        if ($adherent->isHost() || $adherent->isReferent()) {
```

Comment:
This condition seems fragile to me. If another "type" of user is introduced later that is not supposed to be allowed to unregister, it'll still be allowed to do it. At least, this code will have to be updated.

As far as I can see, only "basic adherent" are allowed to unregister. The condition should be simply:

```
return $adherent->isBasicAdherent() ? self::ACCESS_GRANTED : self::ACCESS_DENIED;
```